Implement vibrations

mpnico opened the pull request:
### Implements (non-HD) vibrations in Ryujinx.
Based on previous PRs #1665 and #1436 and adapted for Miria (#2194)

Many thx to @AcK77  for figuring out how to handle and convert correctly the values. This PR aims to bring accurate vibrations to non-HD rumble devices supported by the SDL2 lib.

Add some rumble options in Controller settings :
![image](https://user-images.githubusercontent.com/1097256/125710839-21f55697-0958-46e5-8959-797ab6531543.png)


#### Tested games
- Mario Kart 8 Deluxe
- Super Mario Odyssey
- Super Mario 3D World
- Diablo III
- Monster Hunter Rise

#### Tested controller
- PS5 : some vibrations seems to be skipped, for example in SMO menu but I don't think it is related to something I can fix. More over, the rumble multiplier has very little effect.
- Xbox One
- Xbox 360
- Switch Pro : currently in SDL2, vibrations are only ON/OFF so slider multiplier in UI has no effect.

#### TODO
- [x] Add UI option to enable / disable rumble
- [x] Maybe Add an UI option to provide a multiplier to vibration (left / right) motor
- [x] ~~Dig out better frequency calculation~~ : done through the rework 
- [x] ~~Figure out how to use amplification~~ : done through the rework

Shahil-Ayato commented:
Is this also for HD rumble or just regular rumble for now?

RyzenDew commented:
It works well and feels nice but in smo when rumble should help you find the moon in the ground it wont stop vibrating seas kingdom green pipe
Also tested just dance 2021 and it works great there.

Mou-Ikkai commented:
Game: Super Smash Bros. Ultimate
Controller: Xbox One Wireless (MODEL 1708) w/ Xbox Wireless Adapter.

When going from the intro into the main menu, the rumble doesn't stop naturally. (It took about 1 minute to stop, I used a stopwatch.)
TEST 2: None of SSBU's rumbles stop at the correct time. The logs seem to point that: aLow and aHigh reset to 0, but fLow and fHIgh stay at 160-320 respectively at any time. I believe the stop of the rumbling may be caused by the Xbox adapter/Controller in a timeout of sorts.

[SSBU-RUMBLE.log](https://github.com/Ryujinx/Ryujinx/files/6798082/Ryujinx_1.0.0%2B926bb10_2021-07-11_21-21-43.log)

MutantAura commented:
### Initial tests

**Super Smash Bros. Ultimate** - Controller first vibrates on pressing START at title and never stops. Continuous vibration throughout game and only stops when closing game. Reconnecting controller will stop the effect but it will start as soon as any vibration effect is used again. Worth noting that other vibrations still happened at the correct time e.g. on move hits etc but a background vibration was persistent.

**Mario Kart 8 Deluxe** - Perfect as far as I can tell. Will test more between Ryu and switch later but for now I can't detect any major issues.

**The Legend of Zelda: Breath of the Wild** - Could not notice any vibrations/rumble at all. Unsure of what the issue is here.

**Pokémon Shield** - Vibration works on main menu when not in-game. Need to check this against real hardware as it's been so long since I've played.

Thog commented on Ryujinx.Input/HLE/NpadManager.cs:
```diff
@@ -174,6 +176,15 @@ public void Update(float aspectRatio = 0)
 
                         controller.UpdateUserConfiguration(inputConfig);
                         controller.Update();
+                        if (!_device.Hid.Npads.RumbleQueues.TryGetValue(playerIndex, out ConcurrentQueue<HidVibrationValue> rumbleQueue))
```
imo this if block should be done in the HLE project.

Thog replied:
yes

Thog commented on Ryujinx.Input/HLE/NpadController.cs:
```diff
@@ -492,5 +494,18 @@ public void Dispose()
+                // FrequencyHigh oscillate around stopValues.FrequencyHigh
+                float highFreq = Math.Abs(vibrationValue.FrequencyHigh - stopValues.FrequencyHigh) / stopValues.FrequencyHigh;
+                _gamepad.RumbleInfinity(lowFreq, highFreq);
+                Logger.Info?.Print(LogClass.Hid, "New rumble effect aLow=" + vibrationValue.AmplitudeLow + ", fLow=" + vibrationValue.FrequencyLow + ", aHigh=" + vibrationValue.AmplitudeHigh + ", fHigh=" + vibrationValue.FrequencyHigh);
```
Reminder to remove this before undrafting ^^

Thog commented on Ryujinx.Input/HLE/NpadController.cs:
```diff
@@ -492,5 +494,18 @@ public void Dispose()
+            {
+                // FrequencyLow oscillate around stopValues.FrequencyLow
+                float lowFreq = Math.Abs(vibrationValue.FrequencyLow - stopValues.FrequencyLow) / stopValues.FrequencyLow;
+                // FrequencyHigh oscillate around stopValues.FrequencyHigh
```
```suggestion

                // FrequencyHigh oscillate around stopValues.FrequencyHigh
```

Thog commented on Ryujinx.Input/HLE/NpadController.cs:
```diff
@@ -492,5 +494,18 @@ public void Dispose()
+            if (queue.TryDequeue(out HidVibrationValue vibrationValue))
+            {
+                // FrequencyLow oscillate around stopValues.FrequencyLow
+                float lowFreq = Math.Abs(vibrationValue.FrequencyLow - stopValues.FrequencyLow) / stopValues.FrequencyLow;
```
I think this is a nice middle ground in term of algorithm, need testing but still.

mpnico replied:
That was actually wrong, I use amplitude now

Thog commented on Ryujinx.Input/HLE/NpadController.cs:
```diff
@@ -492,5 +494,18 @@ public void Dispose()
+                float lowFreq = Math.Abs(vibrationValue.FrequencyLow - stopValues.FrequencyLow) / stopValues.FrequencyLow;
+                // FrequencyHigh oscillate around stopValues.FrequencyHigh
+                float highFreq = Math.Abs(vibrationValue.FrequencyHigh - stopValues.FrequencyHigh) / stopValues.FrequencyHigh;
+                _gamepad.RumbleInfinity(lowFreq, highFreq);
```
I do think the system module auto reset to the stop state after a quite short amount of time, will check again later during the week.

mpnico replied:
During my tests on last commit, I was never able to reproduce an infinite rumble situation. But more tests is needed to be sure.

Thog commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -239,50 +244,50 @@ private void SetupNpad(PlayerIndex player, ControllerType type)
             switch (type)
             {
                 case ControllerType.ProController:
-                    controller.StyleSet           = NpadStyleTag.FullKey;
```
All of this is aligned for clarity, rollback all those changes.

Thog commented on Ryujinx.Input/IGamepad.cs:
```diff
@@ -66,7 +66,14 @@ public interface IGamepad : IDisposable
+        /// </summary>
+        /// <param name="lowFrequency">The intensity of the low frequency from 0.0f to 1.0f</param>
+        /// <param name="highFrequency">The intensity of the high frequency from 0.0f to 1.0f</param>
+        void RumbleInfinity(float lowFrequency, float highFrequency);
```
I don't think ``RumbleInfinity`` is needed.
Would be better to just send uint.MaxValue and remap on the SDL2 gamepad side if ``durationMs == uint.MaxValue`` and throw an exception if ``durationMs > SDL_HAPTIC_INFINITY`` otherwise.

Thog commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -596,5 +601,28 @@ private void UpdateDisconnectedInputSixAxis(PlayerIndex index)
+            {
+                if (!LastVibrationValues.ContainsKey(index))
+                {
+                    // first value received seems to be the "stop" values
```
imo we could easily find the "stop" values and use them as constant.
There is also no guaranties that we will receive a "stop" value first.

mpnico replied:
Absolutely true :) That approach was wrong.

Niwu34 commented:
The 8bit do mode of the sn30 plus doesn't seem to work with this Pr
switch/xinput mode work tho
I have a feeling that the Feedback is always a little too long (Need to compare it with my switch later)
Monster Hunter Stories 2 starts with one minute rumble and my controller starts vibrating for a whole min evertime there is Rumble in the game

Benedani commented:
I feel like the long rumbles are being caused by the amplification values being ignored.
Would multiplying the frequency with the amplification work?

mpnico commented:
This is more complex. Currently, SDL2 lib only implements "regular" rumble for now : two motors, "fixed" frequency, variable intensity. So we have to find a "magic formula" to translate HD rumble to regular rumble. Using only amplification seems not enough, using only frequency seems worse

RyzenDew commented:
One of the yuzu prs has an sdl2 git that has hd rumble wip

Mou-Ikkai commented:
> One of the yuzu prs has an sdl2 git that has hd rumble wip

That's... Irrelevant.

CeruleanSky commented:
> One of the yuzu prs has an sdl2 git that has hd rumble wip

https://github.com/german77/SDL/tree/hd_rumble  
The first commit to the tree has some frequency tables if they aren't already present in Ryujinx and the PR mentions that SDL will have HD rumble support in the future so perhaps it will be eventually up-streamed.

mpnico commented:
Updated the PR, much better implementation.

Mou-Ikkai commented:
Game: Super Smash Bros. Ultimate
Controller: Xbox One Wireless (MODEL 1708) w/ Xbox Wireless Adapter.

Rumble seems normal after latest patch. Rumbles seem to match that of the game being run on a native switch. Plus the rumbles don't overextend when there are FPS drops. It also didn't break when I set myself to be hit by multiple enemies at once.

Mou-Ikkai commented:
Game: Project DIVA Mega Mix
Controller: Xbox One Wireless (MODEL 1708) w/ Xbox Wireless Adapter.
Song: Meteor

As a test of simple rumbles, I decided to try the long-Sliders on MegaMix. It was mostly fine, one issue I saw is that sometimes, the rumble would weaken at a random part of the slider, then suddenly come back near the middle or end of the slide.

mpnico commented:
Updated PR with the UI work.

Benedani commented:
Seems to work in my tested games: SMO, SSBU, MK8 Deluxe, 3D World, Super Mario Maker 2, BotW, Terraria and Pokemon Sword (literally the only time you get rumble is in Dynamax battles btw, great game). Minecraft's absence of rumble also works.

emmauss commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -596,5 +608,49 @@ private void UpdateDisconnectedInputSixAxis(PlayerIndex index)
             WriteNewSixInputEntry(ref currentNpad.JoyRightSixAxisSensor, ref newState);
         }
+
+        private Boolean IsSameVibrationValues(HidVibrationValue val1, HidVibrationValue val2)
```
use `bool` alias.
Consider replacing with an override of `Equals` of  `HidVibrationValue`

emmauss commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -25,6 +26,17 @@ public class NpadDevices : BaseDevice
 
+        public Dictionary<PlayerIndex, ConcurrentQueue<(HidVibrationValue, HidVibrationValue)>> RumbleQueues = new Dictionary<PlayerIndex, ConcurrentQueue<(HidVibrationValue, HidVibrationValue)>>();
+        public Dictionary<PlayerIndex, (HidVibrationValue, HidVibrationValue)> _lastVibrationValues = new Dictionary<PlayerIndex, (HidVibrationValue, HidVibrationValue)>();
+        private HidVibrationValue _neutralVibrationValue = new HidVibrationValue
```
Use a static  field to represent the neutral value. It can be used below

emmauss commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -596,5 +608,49 @@ private void UpdateDisconnectedInputSixAxis(PlayerIndex index)
+        {
+            if (!_lastVibrationValues.TryGetValue(index, out (HidVibrationValue, HidVibrationValue) dualVibrationValue))
+            {
+                return new HidVibrationValue
```
can be replaced with static field for neutral value

emmauss commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -596,5 +608,49 @@ private void UpdateDisconnectedInputSixAxis(PlayerIndex index)
+                    FrequencyHigh = 320f
+                };
+            }
+            if (position == 0)
```
This can be simplified with a ternary conditional operator

AcK77 commented on Ryujinx/Configuration/ConfigurationState.cs:
```diff
@@ -859,6 +860,26 @@ public void Load(ConfigurationFileFormat configurationFileFormat, string configu
+            {
+                Common.Logging.Logger.Warning?.Print(LogClass.Application, $"Outdated configuration version {configurationFileFormat.Version}, migrating to version 29.");
+
+                foreach(InputConfig config in configurationFileFormat.InputConfig)
```
```suggestion
                foreach (InputConfig config in configurationFileFormat.InputConfig)
```

AcK77 commented on Ryujinx/Ui/Windows/ControllerWindow.cs:
```diff
@@ -1045,6 +1062,13 @@ private void ProfileLoad_Activated(object sender, EventArgs args)
                             Sensitivity  = 100,
                             GyroDeadzone = 1,
+                        },
+
```
remove extra empty line

AcK77 commented on Ryujinx.Headless.SDL2/Program.cs:
```diff
@@ -236,6 +236,13 @@ private static InputConfig HandlePlayerConfiguration(string inputProfileName, st
                             Sensitivity  = 100,
                             GyroDeadzone = 1,
+                        },
+
```
remove extra empty line

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -596,5 +608,49 @@ private void UpdateDisconnectedInputSixAxis(PlayerIndex index)
+                if (!dualVibrationValues.TryGetValue(0, out HidVibrationValue leftVibrationValue))
+                {
+                    leftVibrationValue = _neutralVibrationValue;
+                }
```
Add a new empty line after

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -596,5 +608,49 @@ private void UpdateDisconnectedInputSixAxis(PlayerIndex index)
+                if (!dualVibrationValues.TryGetValue(1, out HidVibrationValue rightVibrationValue))
+                {
+                    rightVibrationValue = _neutralVibrationValue;
+                }
```
Add a new empty line after

mpnico commented:
Added requested changes and implemented missing GetVibrationDeviceInfo Service, thx to @AcK77

This fix the lack of rumble in Monster Hunter Rise (and probably other games too)

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/IHidServer.cs:
```diff
@@ -1138,13 +1192,27 @@ public ResultCode SendVibrationValues(ServiceCtx context)
+                    {
+                        context.Device.Hid.Npads.UpdateRumbleQueue(currentIndex, dualVibrationValues);
+                        dualVibrationValues = new Dictionary<byte, HidVibrationValue>();
+                    }
```
Add a new extra empty line after

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/ResultCode.cs:
```diff
@@ -7,6 +7,9 @@ enum ResultCode
         Success = 0,
 
-        InvalidNpadIdType = (710 << ErrorCodeShift) | ModuleId
+        InvalidNpadDeviceType = (122 << ErrorCodeShift) | ModuleId,
```
It could be nice to keep the alignment as we did in other `ResultCode.cs` file for considtency

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/ResultCode.cs:
```diff
@@ -7,6 +7,9 @@ enum ResultCode
-        InvalidNpadIdType = (710 << ErrorCodeShift) | ModuleId
+        InvalidNpadDeviceType = (122 << ErrorCodeShift) | ModuleId,
+        InvalidNpadIdType = (123 << ErrorCodeShift) | ModuleId,
+        InvalidDeviceIndex = (124 << ErrorCodeShift) | ModuleId,
```
124 is duplicated

AcK77 commented on Ryujinx.Input.SDL2/SDL2Gamepad.cs:
```diff
@@ -151,7 +151,18 @@ public void Rumble(float lowFrequency, float highFrequency, uint durationMs)
+                }
+                else if (durationMs > SDL_HAPTIC_INFINITY)
+                {
+                    throw new NotSupportedException($"Unsupported rumble duration {durationMs}");
```
I’m not sure we should throw an exception here, maybe a logging error is enough, since it will not stop the execution for a vibration error.

AcK77 commented on Ryujinx.Input/HLE/NpadController.cs:
```diff
@@ -492,5 +495,29 @@ public void Dispose()
+
+        public void UpdateRumble(ConcurrentQueue<(HidVibrationValue, HidVibrationValue)> queue)
+        {
+            if (queue.TryDequeue(out (HidVibrationValue, HidVibrationValue) dualVvibrationValue))
```
```suggestion
            if (queue.TryDequeue(out (HidVibrationValue, HidVibrationValue) dualVibrationValue))
```

AcK77 commented on Ryujinx.Input/HLE/NpadController.cs:
```diff
@@ -492,5 +495,29 @@ public void Dispose()
+
+                    _gamepad.Rumble(low, high, uint.MaxValue);
+
+                    Logger.Debug?.Print(LogClass.Hid, $"New rumble effect for {controllerConfig.PlayerIndex} " +
```
The logger should print the method name `UpdateRumble` so the message text could be shorter.

AcK77 commented on Ryujinx.Input/HLE/NpadManager.cs:
```diff
@@ -174,6 +176,12 @@ public void Update(float aspectRatio = 0)
 
                         controller.UpdateUserConfiguration(inputConfig);
                         controller.Update();
+                        if (!_device.Hid.Npads.RumbleQueues.TryGetValue(playerIndex, out ConcurrentQueue<(HidVibrationValue, HidVibrationValue)> rumbleQueue))
```
Add a new extra empty line before

mpnico replied:
Block has been moved to NpadDevices

AcK77 commented on Ryujinx.Input/HLE/NpadManager.cs:
```diff
@@ -174,6 +176,12 @@ public void Update(float aspectRatio = 0)
+                            rumbleQueue = new ConcurrentQueue<(HidVibrationValue, HidVibrationValue)>();
+                            _device.Hid.Npads.RumbleQueues[playerIndex] = rumbleQueue;
+                        }
+                        controller.UpdateRumble(rumbleQueue);
```
Add a new extra empty line before

mpnico replied:
Block has been moved to NpadDevices

AcK77 commented on Ryujinx.Input/IGamepad.cs:
```diff
@@ -66,7 +66,7 @@ public interface IGamepad : IDisposable
 
         /// <summary>
-        /// Starts a rumble effect on the gampead.
+        /// Starts a rumble effect on the gamepad.
```
Nice catch!

AcK77 reviewed (changes requested):
Just some code style issues for a first batch, I’ll take a closer look when this will be addressed! Thanks for working on the rumble feature ;)

mpnico commented:
Rebased and added requested changes (but one, on ResultCode)

MelonSpeedruns commented:
I've played Skyward Sword for about 2 hours with this PR with an official Pro Controller. There was no bugs or weird issues I've encountered! GG!

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -596,5 +608,48 @@ private void UpdateDisconnectedInputSixAxis(PlayerIndex index)
+        }
+
+        public ConcurrentQueue<(HidVibrationValue, HidVibrationValue)> GetRumbleQueue(PlayerIndex index)
+{
```
Wrong alignment

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:
```diff
@@ -596,5 +608,48 @@ private void UpdateDisconnectedInputSixAxis(PlayerIndex index)
+                rumbleQueue = new ConcurrentQueue<(HidVibrationValue, HidVibrationValue)>();
+                _device.Hid.Npads.RumbleQueues[index] = rumbleQueue;
+            }
+            return rumbleQueue;
```
Add a new empty line before.

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/IHidServer.cs:
```diff
@@ -1138,13 +1185,30 @@ public ResultCode SendVibrationValues(ServiceCtx context)
+                    dualVibrationValues[position] = vibrationValues[deviceCounter];
+                    currentIndex = index;
+                }
+                context.Device.Hid.Npads.UpdateRumbleQueue(currentIndex, dualVibrationValues);
```
Add a new empty line before.

AcK77 commented on Ryujinx.HLE/HOS/Services/Hid/ResultCode.cs:
```diff
@@ -7,6 +7,9 @@ enum ResultCode
+        InvalidNpadDeviceType = (122 << ErrorCodeShift) | ModuleId,
+        InvalidNpadIdType     = (123 << ErrorCodeShift) | ModuleId,
+        InvalidDeviceIndex    = (124 << ErrorCodeShift) | ModuleId,
+        InvalidBufferSize     = (124 << ErrorCodeShift) | ModuleId
```
```suggestion
        InvalidBufferSize     = (131 << ErrorCodeShift) | ModuleId
```
And it should be returned in `SendVibrationValues` call if `vibrationValues.Length != deviceHandles.Length`

mpnico replied:
Thx for the right value !

AcK77 reviewed (changes requested):
Some nits left, after that everything should be good!

mpnico commented:
Adressed last requested changes.

RyzenDew commented:
rumble does not work on the joy cons at all not sure if you can add it or not

mpnico commented:
> rumble does not work on the joy cons at all not sure if you can add it or not

I will try to investigate the problem.

RyzenDew commented:
> > rumble does not work on the joy cons at all not sure if you can add it or not
> 
> 
> 
> I will try to investigate the problem.

Guess it will be in another pr ... lol also thank you
